### ISSUE TITLE
Dealing with clearing and optional fields through a GraphQL API, in a systematic way

### ISSUE THREAD

**Issue by advanderveer** [ent/ent#2145]
We're pretty far along the development of a GraphQL API using Ent. We have nothing but praise for te project (so thank you)! but we're running into a design issue that I'm not sure how to solve in a systematic way. I feels like an obvious thing so we're probably just missing something so hopefully someone is able to help us with this.

Basically we have entities that we want to partially update sometimes. So in one part of the application we would have a form with all the fields being populated by the user, while in other parts it would only have fields to update some parts of the entity. The crux of the issue is that some the fields on the entity we also like to be able to "clear". How can we implement a mutation that is both able to detect that a frontend call wants to explicitly "clear" a field, while also supporting partial updates of the entity being mutated.

On the Graphql/javascript camps it seems to be possible to do this because there is a difference between an explicit "null" and "undefined" (https://medium.com/workflowgen/graphql-mutations-partial-updates-implementation-bff586bda989). But in Go this is not the case. As soon as the value arrives in the Go resolver it will either be a nil or a pointer to the value: the "undefined" information is lost at that point.

I feel like I'm missing something because it seems like an obvious problem so maybe I didn't search well enough. I did find this recent thread (https://github.com/ent/ent/issues/1979) that seems to hint that Ent is more designed with "nil = ignore" in the resolver than "nil = clear". But maybe there is a way to have both?

Thank you in advance!

**Comment by henrywoody** [comment]
You are correct that ent treats `nil` as "ignore" rather than clear with `SetNillable<F>`. To clear a field (that can be cleared) you can use the `Clear<F>` methods. Since Go, unlike JavaScript, makes not distinction between `null` and `undefined`, you can't tell from a single field wether `nil` means "clear" or "don't update", so you'll have to modify your API a bit so that it mirrors ent's API.

Generally, I use a GraphQL schema like this:

```graphql
input MyModelUpdateInput {
    field: Int
    clearField: Boolean
}
```

Then on the server (with ent), you can do something like this:

```go
query := client.MyModel.UpdateOneID(id)

if input.Field != nil {
    query.SetField(*input.Field)
}
// or just: query.SetNillableField(input.Field)

if input.ClearField != nil && *input.ClearField {
    query.ClearField()
}
```

Then it's just up to your client to call the API correctly. If using the API programmatically, this should be fairly straightforward. If you have a user-facing app you can check which fields in the form were modified and create the GraphQL input accordingly. Here's a sample with a JS client:

```
function onSubmit(formValues) {
    const queryInput = {...formValues};
    if (formValues.field === null) {
        queryInput.clearField = true;
    }
    // call your API with `queryInput`
}
```

**Comment by a8m** [comment]
Hey @advanderveer 👋, thanks for the kind words, and thanks @henrywoody for the help.

I'll try to TL;DR this because I plan to publish at some point our GraphQL roadmap with more details. 

As you mentioned, there's no way to distinguish between `nil` and `underfined` if you work with structs (unlike maps), but as @henrywoody mentioned, you can use the `Clear<F>` option. 

Having said that, I plan to support `null` values in GraphQL without defining `clear<F>` in the client-side code, and without using maps on the Go side. We'll do it by reading the mutation input information from the context using the `gqlgen/graphql` package, and then ent will set the `Clear<F>` automatically. The end-users will be able to read `Clear<F>` fields without changing any client-side code.

If you want to use it today, you can do it manually using the package mentioned above, but the current API is not so friendly. e.g. 
```go
// Check if input.text was set explicitly not "null".
f := graphql.GetFieldContext(ctx).Field
nullText := f.Field.Arguments.ForName("input").Value.Children.ForName("text").Raw == "null"
```

**Comment by advanderveer** [comment]
@henrywoody  thank you for the response and providing a workaround! 

@a8m Great to hear you're planning to support it out of the box! About the "not so friendly" solution. Is your plan to include the above check in a template that calls "ClearX" or "SetNillableX" conditionally or are you thinking of something more advanced. If it is the former our team might be able to contribute this to the codebase.

**Comment by a8m** [comment]
The plan is to provide official support for mutation inputs. Today, we provide a template example on how to generate such inputs using ent templates (see [a8m/ent-graphql-example/mutation_input.tmpl](https://github.com/a8m/ent-graphql-example/blob/master/ent/template/mutation_input.tmpl)). We actually use a version of this template ourselves. 

If you're using this template, you're welcome to contribute a change to it. If not, we'll post soon our roadmap and the new API for entgql, and we'll love to get more help there 😃
